File: src/main/java/org/jukeboxmc/inventory/BarrelInventory.java
package org.jukeboxmc.inventory;

import com.nukkitx.protocol.bedrock.data.SoundEvent;
import com.nukkitx.protocol.bedrock.data.inventory.ContainerType;
import org.jetbrains.annotations.NotNull;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.block.behavior.BlockBarrel;
import org.jukeboxmc.blockentity.BlockEntityBarrel;
import org.jukeboxmc.player.Player;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BarrelInventory extends ContainerInventory {

    public BarrelInventory( InventoryHolder holder ) {
        super( holder, -1, 27 );
    }

    @Override
    public BlockEntityBarrel getInventoryHolder() {
        return (BlockEntityBarrel) this.holder;
    }

    @Override
    public @NotNull InventoryType getType() {
        return InventoryType.BARREL;
    }

    @Override
    public @NotNull ContainerType getWindowTypeId() {
        return ContainerType.CONTAINER;
    }

    @Override
    public void onOpen(@NotNull Player player ) {
        super.onOpen( player );
        if ( this.viewer.size() == 1 ) {
            Block block = this.getInventoryHolder().getBlock();
            if ( block instanceof BlockBarrel blockBarrel ) {
                if ( !blockBarrel.isOpen() ) {
                    blockBarrel.setOpen( true );
                    player.getWorld().playSound( block.getLocation(), SoundEvent.BARREL_OPEN );
                }
            }
        }
    }

    @Override
    public void onClose(@NotNull Player player ) {
        if ( this.viewer.size() == 0 ) {
            Block block = this.getInventoryHolder().getBlock();
            if ( block instanceof BlockBarrel blockBarrel ) {
                if ( blockBarrel.isOpen() ) {
                    blockBarrel.setOpen( false );
                    player.getWorld().playSound( blockBarrel.getLocation(), SoundEvent.BARREL_CLOSE );
                }
            }
        }
    }
}

File: src/main/java/org/jukeboxmc/inventory/BarrelInventory.kt
package org.jukeboxmc.inventory

import org.cloudburstmc.protocol.bedrock.data.SoundEvent
import org.cloudburstmc.protocol.bedrock.data.inventory.ContainerType
import org.jukeboxmc.block.Block
import org.jukeboxmc.block.behavior.BlockBarrel
import org.jukeboxmc.blockentity.BlockEntityBarrel
import org.jukeboxmc.player.Player

/**
 * @author LucGamesYT
 * @version 1.0
 */
class BarrelInventory(holder: InventoryHolder?) : ContainerInventory(holder, -1, 27) {
    override val inventoryHolder: BlockEntityBarrel
        get() = holder as BlockEntityBarrel
    override val type: InventoryType
        get() = InventoryType.BARREL
    override val windowTypeId: ContainerType
        get() = ContainerType.CONTAINER

    override fun onOpen(player: Player) {
        super.onOpen(player)
        if (viewer.size == 1) {
            val block: Block = inventoryHolder.block
            if (block is BlockBarrel) {
                if (!block.isOpen) {
                    block.isOpen = true
                    player.world?.playSound(block.location, SoundEvent.BARREL_OPEN)
                }
            }
        }
    }

    override fun onClose(player: Player) {
        if (viewer.size == 0) {
            val block: Block = inventoryHolder.block
            if (block is BlockBarrel) {
                if (block.isOpen) {
                    block.isOpen = false
                    player.world?.playSound(block.location, SoundEvent.BARREL_CLOSE)
                }
            }
        }
    }
}
